Add concurrency policy

thomasjhuang opened the pull request:
## Tracking issue
This is associated to flyte change https://github.com/flyteorg/flyte/pull/6475, and RFC https://github.com/flyteorg/flyte/pull/5659 

## Why are the changes needed?
These are the necessary changes needed to implement the ConcurrencyPolicy within flytekit. Namely, it will allow users to manage concurrency control via `LaunchPlan.create()` or `LaunchPlan.get_or_create()`

## How was this patch tested?

Unit tests are added, local testing is also done.

### Screenshots

## Check all the applicable boxes <!-- Follow the above conventions to check the box -->

- [ ] I updated the documentation accordingly.
- [ ] All new and existing tests passed.
- [ ] All commits are signed-off.

## Related PRs

<!-- Add related pull requests for reviewers to check -->

## Docs link

<!-- Add documentation link built by CI jobs here, and specify the changed place --> 
 <div id='description'>
    <a href="https://bito.ai#summarystart"></a>
<h3>Summary by Bito</h3>
This pull request introduces a new concurrency policy feature in the Flytekit library, enhancing concurrency control through LaunchPlan methods. It updates existing models and launch plan parameters to incorporate these settings, improving overall functionality. Comprehensive unit tests have been added to ensure reliability and system integrity.
<!-- Disabling unit_tests and post_effort_to_review
<br>
<br>
<b>Unit tests added</b>: True
<br>
<br>
<b>Estimated effort to review (1-5, lower is better)</b>: 2
-->
</div>

Sovietaced commented:
@thomasjhuang Can you sign off on these commits? And can someone else let the CI run?

thomasjhuang commented:
> @thomasjhuang Can you sign off on these commits? And can someone else let the CI run?

Sorry about that, added the signoff

thomasjhuang commented on flytekit/models/concurrency.py:
```diff
@@ -0,0 +1,61 @@
+
+
+class ConcurrencyLimitBehavior(object):
+    SKIP = _launch_plan_idl.ConcurrencyLimitBehavior.SKIP
```
Yea I need to update this according to the merged PR in flyte. Thanks for the callout

kumare3 commented:
LGTM

thomasjhuang commented on flytekit/models/concurrency.py:
```diff
@@ -0,0 +1,62 @@
+from flyteidl.admin import launch_plan_pb2 as _launch_plan_idl
```
Actually if we look at other related files within `/models` the references aren't different for example dynamic_job does this `from flyteidl.core import dynamic_job_pb2 as _dynamic_job`. I believe the reference I made was wrong, but I'm not entirely sure. Can you help run the checks again?

thomasjhuang commented:
@kumare3 @davidmirror-ops These test failures are a little odd to me, they seem transient? would a rerun suffice here?

kumare3 commented:
Hmm weird how did the event loop vanish

thomasjhuang commented:
> I fixed the failed test in https://github.com/flyteorg/flytekit/pull/3265/files#r2354426173 @thomasjhuang could you please merge/rebase master and see if tests can pass? Thank you!

Ah got it, thanks I've rebased and pushed again.

machichima commented on tests/flytekit/unit/models/test_concurrency.py:
```diff
@@ -0,0 +1,41 @@
+
+
+def test_concurrency_limit_behavior():
+    assert ConcurrencyLimitBehavior.SKIP == _launch_plan_idl.ConcurrencyLimitBehavior.SKIP
```
```suggestion
    assert ConcurrencyLimitBehavior.SKIP == _launch_plan_idl.CONCURRENCY_LIMIT_BEHAVIOR_SKIP
```

All `_launch_plan_idl.ConcurrencyLimitBehavior.SKIP` below should be changed to `_launch_plan_idl.CONCURRENCY_LIMIT_BEHAVIOR_SKIP`, then the test should pass

thomasjhuang commented:
@kumare3 @davidmirror-ops @popojk Looks like all checks have passed now!

machichima commented on flytekit/core/launch_plan.py:
```diff
@@ -226,6 +230,7 @@ def get_or_create(
         :param trigger: [alpha] This is a new syntax for specifying schedules.
         :param overwrite_cache: If set to True, the execution will always overwrite cache
         :param auto_activate: If set to True, the launch plan will be activated automatically on registration.
+        :param concurrency: Defines execution concurrency limits and policy when limit is reached
          Default is False.
```
Is the docstring here being cut-off?

machichima replied:
```suggestion
        :param auto_activate: If set to True, the launch plan will be activated automatically on registration.
            Default is False.
        :param concurrency: Defines execution concurrency limits and policy when limit is reached.
```

Oh I mean "Default is False" is for `auto_activate`?

machichima commented on tests/flytekit/unit/models/test_concurrency.py:
```diff
@@ -0,0 +1,41 @@
+    assert ConcurrencyLimitBehavior.enum_to_string(999) == "<UNKNOWN>"
+
+
+def test_concurrency_policy():
```
```suggestion
def test_concurrency_policy_serialization():
```
nit: making test name more clear
